File: alpha-beta/engine.py
import time
import multiprocessing
from functools import partial

import chess
import eval

def genMoveIterative(seconds: float, board: chess.Board) -> str:    
    timeout = time.time() + seconds
    depth = 2
    move = None
    while time.time() < timeout:
        depth += 1
        move = genMove(depth, board)
    print(f"depth reached - {depth}")
    return move

# generate move with fixed depth alpha-beta search
def genMove(depth: int, board: chess.Board) -> str:
    moves = getOrderedMoves(board=board)
    if not moves:
        return None, 0
    if board.is_checkmate():
        return None, float('-inf')
    
    pool = multiprocessing.Pool()
    eval = partial(evaluate_move, depth, board.copy())
    results = pool.map(eval, moves)
    pool.close()
    pool.join()

    bestValue = float('-inf')
    bestMove = None
    for move, value in results:
        if value >= bestValue:
            bestValue = value
            bestMove = move
    return bestMove, bestValue

def evaluate_move(depth: int, board: chess.Board, move: chess.Move) -> tuple:
    board.push(move)
    value = -negaMax(depth - 1, float('-inf'), float('inf'), board)
    board.pop()
    return move, value

def negaMax(depth: int, alpha: float, beta: float, board: chess.Board) -> float:
    if board.is_checkmate():
        return -100000 * (depth + 1)
    if board.can_claim_draw():
        return 0
    moves = board.legal_moves
    if not moves:
        return 0
    if depth == 0:
        return qSearch(alpha, beta, board)
    
    for move in moves:
        board.push(move)
        value = -negaMax(depth - 1, -beta, -alpha, board)
        board.pop()
        if value >= beta:
            return beta
        alpha = max(alpha, value)
    
    return alpha
    
def qSearch(alpha: float, beta: float, board: chess.Board):
    value = eval.evaluate(board)
    if value >= beta:
        return beta
    alpha = max(alpha, value)
    moves = board.legal_moves
    for move in moves:
        if not board.is_capture(move):
            continue
        board.push(move)
        value = -qSearch( -beta, -alpha, board)
        board.pop()
        if value >= beta:
            return beta
        alpha = max(alpha, value)
    return alpha

def getOrderedMoves(board: chess.Board):
    return sorted(list(board.legal_moves), key=lambda move: eval.evaluateMove(move, board))

File: alpha-beta-go/engine.go
package main

import (
	"math"
	"time"

	"github.com/corentings/chess"
)

const (
	mateScore     = 100000.0
	initialDepth  = 2
	maxSearchTime = 5.0
)

var endGame bool

func GenMoveIterative(seconds float64, pos *chess.Position) (*chess.Move, float64) {
	timeout := time.Now().Add(time.Duration(seconds * float64(time.Second))).UnixNano()
	depth := initialDepth
	var move *chess.Move
	var value float64
	for time.Now().UnixNano() < timeout {
		depth++
		move, value = GenMove(depth, pos)
	}
	return move, value
}

func GenMove(depth int, pos *chess.Position) (*chess.Move, float64) {
	moves := getOrderedMoves(pos)
	if len(moves) == 0 {
		return nil, 0
	}

	bestValue := -math.MaxFloat64
	var bestMove *chess.Move
	for _, mv := range moves {
		child := pos.Update(mv)
		value := -negamax(depth-1, -math.MaxFloat64, math.MaxFloat64, child)
		if value > bestValue {
			bestValue = value
			bestMove = mv
		}
	}
	return bestMove, bestValue
}

func negamax(depth int, alpha, beta float64, pos *chess.Position) float64 {
	if pos == nil {
		return 0
	}
	if isCheckmate(pos) {
		return -mateScore * float64(depth+1)
	}
	if canClaimDraw(pos) {
		return 0
	}

	moves := pos.ValidMoves()
	if len(moves) == 0 {
		return 0
	}
	if depth == 0 {
		return quiescenceSearch(alpha, beta, pos)
	}

	for _, mv := range moves {
		child := pos.Update(mv)
		value := -negamax(depth-1, -beta, -alpha, child)
		if value >= beta {
			return beta
		}
		if value > alpha {
			alpha = value
		}
	}
	return alpha
}

func quiescenceSearch(alpha, beta float64, pos *chess.Position) float64 {
	value := evaluate(pos)
	if value >= beta {
		return beta
	}
	if value > alpha {
		alpha = value
	}

	for _, mv := range pos.ValidMoves() {
		if !isCapture(mv, pos) {
			continue
		}
		child := pos.Update(mv)
		value = -quiescenceSearch(-beta, -alpha, child)
		if value >= beta {
			return beta
		}
		if value > alpha {
			alpha = value
		}
	}
	return alpha
}

func getOrderedMoves(pos *chess.Position) []*chess.Move {
	moves := pos.ValidMoves()
	ordered := make([]*chess.Move, len(moves))
	copy(ordered, moves)
	for i := 0; i < len(ordered); i++ {
		for j := i + 1; j < len(ordered); j++ {
			if evaluateMove(ordered[j], pos) > evaluateMove(ordered[i], pos) {
				ordered[i], ordered[j] = ordered[j], ordered[i]
			}
		}
	}
	return ordered
}
